### ISSUE TITLE
Inconsistent casting of floats with jit compilation.

### ISSUE THREAD

**Issue by sschoenholz** [google/jax#732]
```
def mul(R, c):
  return R * c

R = np.array([1, 2, 3], dtype=np.float32)
c = 0.1

print(mul(R, c).dtype) # Evaluates to np.float32 as expected.
print(jit(mul)(R, c).dtype) # Evaluates to np.float64.
```

**Comment by mattjj** [comment]
Thanks for catching this.

I verified that this only happens with jax_enable_x64 switched on. I believe this is a Python scalars issue (which our tests don't cover).

**Comment by mattjj** [comment]
The problem is that [we follow NumPy's special treatment of builtin numeric types in our type promotion logic](https://github.com/google/jax/blob/bb5cbaabaab033a78107987309161041cfa1780b/jax/numpy/lax_numpy.py#L147-L162), but that only works for literals and not argument values passed to`@jit`  functions because we don't have a special abstract value for Python scalars and thus we lose the information we'd need to give them special treatment.

There are at least two things we could want here:
1. follow NumPy
2. don't change behavior when applying `@jit`

I think the latter is more important than the first, and moreover that a heavyweight solution to achieve both probably isn't worth the added complexity.

My current understanding of this issue is that it shows the code is currently achieving some cases of (1) at the expense of (2). We should consider re-prioritizing.

In the meantime, one thing you can do is just not pass Python scalars as arguments. That's unfortunately made slightly trickier by #121, since you can't just write `c = np.array(0.1)` here; you have to write `c = onp.array(0.1)` until we fix #121.
